File: packages/swift/Sources/FrickSwift/FrickProjectionStore.swift
import Foundation
import Observation

// MARK: - FrickProjectionEventSource

/// The slice of ``FrickSyncSocket`` a ``FrickProjectionStore`` depends on.
/// Abstracted so tests can drive a store with hand-fed events and assert the
/// observable row map updates, without standing up a real socket or server.
///
/// Mirrors ``FrickStoreEventSource`` but targets projection subscriptions:
/// `subscribeProjection(name:)` registers the subscription (the socket replays
/// it on reconnect) and the server streams ``FrickInboundEvent/projectionDelta(projection:changes:)``
/// frames the store folds into its keyed cache.
public protocol FrickProjectionEventSource: Sendable {
    /// The inbound event stream the store folds into its row map.
    var events: AsyncThrowingStream<FrickInboundEvent, Error> { get async }

    /// Subscribe to a projection by name. The server streams `ProjectionDelta`
    /// frames carrying keyed row upserts (and deletions via `nil` values).
    func subscribeProjection(name: String) async throws
}

extension FrickSyncSocket: FrickProjectionEventSource {}

// MARK: - FrickProjectionStore

/// A generic, reusable observable projection cache + sync loop.
///
/// `FrickProjectionStore` subscribes to a single projection over the sync
/// socket and maintains an in-memory map keyed by the projection-defined row
/// key, exposing it as observable state for SwiftUI. It mirrors the
/// TypeScript runtime's `client.projection(name)` abstraction, which returns a
/// Signal of `Map<rowKey, row>` fed by `ProjectionDelta` frames.
///
/// It applies, in order, every ``FrickInboundEvent/projectionDelta(projection:changes:)``
/// for its projection name:
///
/// - a change with a non-`nil` value **upserts** the row at `key`,
/// - a change with a `nil` value **deletes** the row at `key`.
///
/// On reconnect the underlying socket replays its registered subscriptions, so
/// a dropped connection self-heals: a fresh snapshot re-arrives and reconciles
/// the map without app intervention. (Identical semantics to the TS runtime,
/// which re-issues projection subscribes after reconnect.)
///
/// ```swift
/// @MainActor
/// final class InboxProjection {
///     let store: FrickProjectionStore
///     init(socket: FrickSyncSocket) {
///         store = FrickProjectionStore(source: socket, name: "conversation-inbox")
///         store.start()
///     }
/// }
/// ```
@MainActor
@Observable
public final class FrickProjectionStore {
    /// The current projection rows, keyed by the projection-defined row key.
    /// Observable: SwiftUI views reading this re-render on every applied delta.
    /// Values are carried as ``FrickMsgPackValue``; use `.anyValue` for a
    /// JSON-friendly Swift tree (mirrors ``FrickProjectionChange``).
    public private(set) var rows: [String: FrickMsgPackValue] = [:]

    /// `true` once at least one ``FrickInboundEvent/projectionDelta(projection:changes:)``
    /// has been applied. Use it to gate a loading spinner vs. an empty-state
    /// view (empty + bootstrapped ≠ loading).
    public private(set) var hasBootstrapped: Bool = false

    /// The last error surfaced by the sync loop, if any. The loop keeps running
    /// (it self-heals on reconnect); this is for diagnostics/UI.
    public private(set) var lastError: String?

    /// The projection name this store subscribes to.
    public let name: String

    @ObservationIgnored private let source: FrickProjectionEventSource
    @ObservationIgnored private var listenerTask: Task<Void, Never>?

    public init(source: FrickProjectionEventSource, name: String) {
        self.source = source
        self.name = name
    }

    deinit {
        listenerTask?.cancel()
    }

    // MARK: Lifecycle

    /// Subscribe to the projection and begin folding inbound deltas into the
    /// row map. Idempotent — a second call while already running is a no-op, so
    /// it's safe to call from `task {}`/`onAppear`.
    public func start() {
        guard listenerTask == nil else { return }
        listenerTask = Task { [weak self] in
            await self?.run()
        }
    }

    /// Tear down on sign-out: cancel the listener and clear the cache. After
    /// `reset()` the store can be `start()`ed again (e.g. on the next sign-in).
    public func reset() {
        listenerTask?.cancel()
        listenerTask = nil
        rows = [:]
        hasBootstrapped = false
        lastError = nil
    }

    // MARK: Sync loop

    private func run() async {
        do {
            try await source.subscribeProjection(name: name)
            for try await event in await source.events {
                switch event {
                case let .projectionDelta(projection, changes) where projection == name:
                    apply(changes: changes)
                case let .status(status):
                    if let err = status.lastError { lastError = err }
                    // On a fresh (re)connect the socket replays the projection
                    // subscription itself; the snapshot re-arrives and
                    // reconciles the map — nothing to do here.
                default:
                    continue
                }
            }
        } catch is CancellationError {
            // Expected on reset()/deinit — swallow.
        } catch {
            // The socket emits its own status updates and reconnects; surface
            // the error for UI but keep the (now-finished) loop from crashing.
            lastError = "\(error)"
        }
    }

    // MARK: Apply

    /// Fold a batch of projection changes into the keyed map. A `nil` value
    /// deletes the row at `key`; a non-`nil` value upserts it. Mirrors the TS
    /// runtime's ProjectionDelta apply (null → delete). Sets
    /// ``hasBootstrapped`` once any delta lands.
    private func apply(changes: [FrickProjectionChange]) {
        for change in changes {
            if let value = change.value {
                rows[change.key] = value
            } else {
                rows.removeValue(forKey: change.key)
            }
        }
        hasBootstrapped = true
    }
}

File: apps/android/frick/src/main/java/dev/frick/client/FrickProjectionStore.kt
package dev.frick.client

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * A generic, reusable observable projection store + sync loop for the Android
 * SDK.
 *
 * Subscribes to a single projection `name` over an existing [FrickSyncSocket],
 * maintains an in-memory map keyed by the projection-defined row key, and
 * exposes the rows as a [StateFlow] for Compose `collectAsState`. It mirrors
 * the TypeScript runtime's `client.projection(name)` abstraction (a Signal of
 * `Map<rowKey, row>` fed by `ProjectionDelta` frames) and the Swift
 * `FrickProjectionStore`.
 *
 * Lifecycle:
 *  - [start] subscribes to the projection and begins collecting
 *    [FrickSyncSocket.events].
 *  - Each [FrickInboundEvent.ProjectionDelta] for [name] is folded into the
 *    map: a change with a non-null [ProjectionChange.value] upserts the row at
 *    its key; a null value deletes it (matching the TS null-deletes semantics).
 *  - On reconnect (the socket transitions back to Ready after a drop) the store
 *    re-subscribes automatically so the server replays a fresh snapshot. The
 *    first projection delta after each (re)subscribe is treated as a snapshot
 *    boundary: the cache is reconciled against the delta's key set so rows the
 *    server dropped while the client was offline (absent from the snapshot) are
 *    pruned instead of lingering as stale rows. Live deltas after the boundary
 *    keep folding in incrementally.
 *
 * Row values are surfaced as the decoded `Map<String, Any?>` payload exactly as
 * delivered on the wire, so callers project them into their own model types.
 */
class FrickProjectionStore(
    private val socket: FrickSyncSocket,
    /** The projection name to subscribe to (matches `subscribeProjection(name)`). */
    private val name: String,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default),
) : AutoCloseable {

    // Insertion-ordered cache keyed by row key. Guarded by `lock` so the event
    // collector and any external apply* calls don't race.
    private val cache = LinkedHashMap<String, Map<String, Any?>>()
    private val lock = Any()

    private val _rows = MutableStateFlow<Map<String, Map<String, Any?>>>(emptyMap())

    /** The live projection rows. Collect with Compose `collectAsState()`. */
    val rows: StateFlow<Map<String, Map<String, Any?>>> = _rows.asStateFlow()

    @Volatile private var collectJob: Job? = null
    @Volatile private var statusJob: Job? = null
    @Volatile private var started = false
    @Volatile private var lastReady = false
    @Volatile private var closed = false

    // Armed on each (re)subscribe; the next projection delta is the authoritative
    // snapshot and the cache is reconciled against it. Guarded by `lock`.
    private var pendingSnapshotReset = false

    /**
     * Begin syncing: subscribe to [name] and start applying deltas. Idempotent —
     * a second call while already running is a no-op.
     */
    fun start() {
        if (closed) return
        synchronized(lock) {
            if (started) return
            started = true
        }
        collectJob = scope.launch {
            socket.events.collect { event -> onEvent(event) }
        }
        // Drive (re)subscription off the socket's Ready transitions so a
        // reconnect replays the snapshot for this projection.
        statusJob = scope.launch {
            socket.status.collect { status ->
                val ready = status is FrickSyncStatus.Ready
                if (ready && !lastReady) {
                    synchronized(lock) { pendingSnapshotReset = true }
                    socket.subscribeProjection(name)
                }
                lastReady = ready
            }
        }
        // Ensure the socket is connecting; harmless if already in flight.
        socket.connect()
    }

    /**
     * Apply a single inbound event to the cache. Public-but-internal seam used
     * by the live collector ([start]) and exercised directly by unit tests that
     * inject [FrickInboundEvent.ProjectionDelta]s without standing up a socket.
     */
    internal fun onEvent(event: FrickInboundEvent) {
        when (event) {
            is FrickInboundEvent.ProjectionDelta -> applyDelta(event)
            else -> Unit
        }
    }

    private fun applyDelta(delta: FrickInboundEvent.ProjectionDelta) {
        if (delta.projection != name) return
        var changed = false
        synchronized(lock) {
            // Snapshot-boundary reconcile on the first delta after a (re)subscribe:
            // prune any cached key the authoritative snapshot does not carry (rows
            // the server dropped while we were offline). Only keys with a non-null
            // value count as present; an explicit null is still a deletion.
            if (pendingSnapshotReset) {
                pendingSnapshotReset = false
                val present = delta.changes
                    .filter { it.value != null }
                    .mapTo(HashSet()) { it.key }
                val staleIterator = cache.keys.iterator()
                while (staleIterator.hasNext()) {
                    if (staleIterator.next() !in present) {
                        staleIterator.remove()
                        changed = true
                    }
                }
            }
            for (change in delta.changes) {
                val value = change.value
                if (value == null) {
                    if (cache.remove(change.key) != null) changed = true
                } else {
                    cache[change.key] = value
                    changed = true
                }
            }
        }
        if (changed) publish()
    }

    /**
     * Test/integration seam: arm the snapshot-boundary reconcile so the next
     * delta is treated as an authoritative snapshot (stale rows pruned). The live
     * path arms this automatically on every (re)subscribe.
     */
    internal fun markResubscribed() {
        synchronized(lock) { pendingSnapshotReset = true }
    }

    /** Current row payload for [key], or null if absent. */
    fun get(key: String): Map<String, Any?>? = synchronized(lock) { cache[key] }

    private fun publish() {
        val snapshot = synchronized(lock) { LinkedHashMap(cache) }
        _rows.value = snapshot
    }

    override fun close() {
        if (closed) return
        closed = true
        collectJob?.cancel()
        statusJob?.cancel()
        scope.cancel()
    }
}
